core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala

```diff
@@ -106,18 +107,30 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
     val taCtxt = getTaskContext()
     val cmtr = getOutputCommitter()
     if (cmtr.needsTaskCommit(taCtxt)) {
-      try {
-        cmtr.commitTask(taCtxt)
-        logInfo (taID + ": Committed")
-      } catch {
-        case e: IOException => {
-          logError("Error committing the output of task: " + taID.value, e)
-          cmtr.abortTask(taCtxt)
-          throw e
+      val outputCommitCoordinator = SparkEnv.get.outputCommitCoordinator
+      val canCommit = outputCommitCoordinator.canCommit(jobID,
+        taID.value.getTaskID().getId(), splitID, attemptID)
+      if (canCommit) {
+        try {
+          cmtr.commitTask(taCtxt)
+          logInfo (s"$taID: Committed")
+        } catch {
+          case e: IOException => {
+            logError("Error committing the output of task: " + taID.value, e)
+            cmtr.abortTask(taCtxt)
+            throw e
+          }
         }
+      } else {
+        val msg: String = s"$taID: Not committed because the driver did not authorize commit"
+        logInfo(msg)
+        throw new CommitDeniedException(msg, jobID, splitID, attemptID)
       }
     } else {
-      logInfo ("No need to commit output of task: " + taID.value)
+      val msg: String = s"No need to commit output of task" +
+        " because needsTaskCommit=false: ${taID.value}"
+      logInfo(msg)
+      throw new CommitDeniedException(msg, jobID, splitID, attemptID)
```

Comment:
My understanding is that if `needsTaskCommit` is false, then the task's output has been successfully committed and therefore we don't ever need to throw an exception from this branch.  However, we still need to throw an exception in the case where the commit is denied by the driver.